piptools/utils.py

```diff
@@ -72,7 +72,10 @@ def key_from_req(req: InstallRequirement | Distribution | Requirement) -> str:
         key = req.key
     else:
         # from packaging, such as install requirements from requirements.txt
-        key = req.name
+        if hasattr(req, "canonical_name"):
```

Comment:
Where should that attribute come from?

Reply:
This comes from pip's internal classes derived from `pip._internal.metadata.BaseDistribution`. Previously, `importlib.metadata.Distribution` or its pkg_resources equivalent was being used. Now added a comment.

With this change, we do start relying on another internal implementation detail of pip, but this package seems to be coupled with the internals enough already, so I thought it was OK. The alternative would be to roll out something like `pip._internal.metadata.BaseDistribution.direct_url` ourselves.

Reply:
I can also have the exact behavior as before by using `._dist.name` here. Let me know.

Reply:
> I don't think this change should be backwards incompatible in any way now.

Well, the behaviour of sync watching the `#egg=` fragments has been dropped...

Reply:
I think that'll be alright.